blockchain/v2/scheduler.go

```diff
@@ -383,3 +469,107 @@ func (sc *schedule) numBlockInState(targetState blockState) uint32 {
 	}
 	return num
 }
+
+// Handlers
+
+// XXX: This handler will probably get the block first and then pass it on to the processor
+func (sc *scheduler) handleBlockResponse(event tmpBlockResponse) (Event, error) {
+	err := sc.touchPeer(event.peerID, event.time)
+	if err != nil {
+		return scPeerError{peerID: event.peerID, reason: err}, nil
+	}
+
+	err = sc.markReceived(event.peerID, event.block.Height, event.size, event.time)
+	if err != nil {
+		return scPeerError{peerID: event.peerID, reason: err}, nil
+	}
+
+	return scBlockReceived{peerID: event.peerID, block: event.block}, nil
+}
+
+func (sc *scheduler) handleBlockProcessed(event pcBlockProcessed) (Event, error) {
+	err := sc.markProcessed(event.height)
+	if err != nil {
+		return scSchedulerFail{reason: err}, nil // this should be fatal
+	}
+
+	if sc.allBlocksProcessed() {
+		return scFinishedEv{}, nil
+	}
+
+	return noOp, nil
+}
+
+// XXX: unify types peerError
+func (sc *scheduler) handlePeerError(event peerError) (Event, error) {
+	err := sc.removePeer(event.peerID)
+	if err != nil {
+		return scSchedulerFail{reason: err}, nil // xxx: bit extreme
+	}
+
+	return noOp, nil
+}
+
+func (sc *scheduler) handleTryPrunePeer(event tryPrunePeer) (Event, error) {
+	prunablePeers := sc.prunablePeers(sc.peerTimeout, sc.minRecvRate, event.time)
+
+	if len(prunablePeers) > 0 {
+		peerID := prunablePeers[0]
+		err := sc.removePeer(peerID)
+		if err != nil {
+			return scSchedulerFail{reason: err}, nil // xxx: should be fatal
+		}
+		return scPeerPruned{peerID: peerID}, nil
+	}
+	return noOp, nil
+}
+
+func (sc *scheduler) handleTrySchedule(event trySchedule) (Event, error) {
+	pendingBlocks := sc.numBlockInState(blockStatePending)
+	receivedBlocks := sc.numBlockInState(blockStateReceived)
+	todo := math.Min(float64(sc.targetPending-pendingBlocks), float64(sc.targetReceived-receivedBlocks))
```

Comment:
I think it's better to have a single `sc.maxBlockPoolSize` to indicate the total number of blocks either pending or received and not processed, and then have something like:
```suggestion
	todo := sc.maxBlockPoolSize-pendingBlocks-receivedBlocks
```